quinn-proto/src/packet.rs

```diff
@@ -25,13 +25,58 @@ pub struct PartialDecode {
 }
 
 impl PartialDecode {
-    pub fn new(bytes: BytesMut, local_cid_len: usize) -> Result<Self, PacketDecodeError> {
+    pub fn new(
+        bytes: BytesMut,
+        local_cid_len: usize,
+    ) -> Result<(Self, Option<BytesMut>), PacketDecodeError> {
         let mut buf = io::Cursor::new(bytes);
         let invariant_header = InvariantHeader::decode(&mut buf, local_cid_len)?;
-        Ok(Self {
-            invariant_header,
-            buf,
-        })
+        let end = match invariant_header {
```

Comment:
So clearly all the other parts of this commit are great -- I think you're correct that what I had was not the right place to put that packet splitting.

However, this hunk makes me uncomfortable. I'm thinking this code should be inlined into `InvariantHeader::decode()`, maybe, or there could be a method on `InvariantHeader` that returns the a `Result<usize, PacketDecodeHeader>` with the payload length?

Reply:
The length definitely isn't part of the invariant header--we're parsing the data that follows it according to QUIC 1.0 assumptions. I'll try to factor it out into a separate function for clarity, but I don't see a way to reuse the same code elsewhere.